libcudacxx/include/cuda/std/__cccl/exceptions.h

```diff
@@ -32,4 +32,54 @@
 #  define _CCCL_HAS_EXCEPTIONS() __EXCEPTIONS
 #endif // has exceptions
 
+// The following macros are used to conditionally compile exception handling code. They
+// are used in the same way as `try` and `catch` blocks, but they allow for different
+// behavior based on whether exceptions are enabled or not, and whether the code is being
+// compiled for device or not.
+//
+// Usage:
+//   _CCCL_TRY({
+//     ...                             // Code that may throw an exception
+//   })
+//   _CCCL_CATCH_OR((cuda_error& e) {  // Handle CUDA exceptions
+//     ...
+//   })
+//   _CCCL_CATCH((...) {               // Handle any other exceptions
+//     ...
+//   })
+#if !_CCCL_HAS_EXCEPTIONS() || (_CCCL_CUDA_COMPILATION() && defined(__CUDA_ARCH__))
```

Comment:
```suggestion
#if !_CCCL_HAS_EXCEPTIONS() || (_CCCL_DEVICE_COMPILATION() && !_CCCL_CUDA_COMPILER(NVHPC))
```
Let's use our macros instead of `__CUDA_ARCH__`

Reply:
is `_CCCL_DEVICE_COMPILATION() == 1 && _CCCL_CUDA_COMPILER(NVHPC) == 1` ever true? what does "device compilation" even mean in nvc++ outside of a function?